src/Workspaces/Core/Portable/Shared/Utilities/StreamingProgressTracker.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Shared.Utilities
+{
+    /// <summary>
+    /// Utility class that can be used to track the progress of an operation in a threadsafe manner.
+    /// </summary>
+    internal class StreamingProgressTracker : IStreamingProgressTracker
+    {
+        private int _completedItems;
+        private int _totalItems;
+
+        private readonly Func<int, int, Task> _updateActionOpt;
+
+        public StreamingProgressTracker()
+            : this(null)
+        {
+        }
+
+        public StreamingProgressTracker(Func<int, int, Task> updateActionOpt)
+        {
+            _updateActionOpt = updateActionOpt;
+        }
+
+        public int CompletedItems => _completedItems;
+
+        public int TotalItems => _totalItems;
+
+        public Task AddItemsAsync(int count)
+        {
+            Interlocked.Add(ref _totalItems, count);
+            return UpdateAsync();
+        }
+
+        public Task ItemCompletedAsync()
+        {
+            Interlocked.Increment(ref _completedItems);
+            return UpdateAsync();
+        }
+
+        public Task ClearAsync()
```

Comment:
When would you foresee this being used?  It doesn't seem to be currently, and it doesn't have the same threadsafety as other member.  Consider removing.